src/SBBox.cpp

```diff
@@ -283,9 +327,238 @@ namespace galsim {
+        SBProfile(new SBTopHatImpl(radius,flux,gsparams)) {}
+
+    SBTopHat::SBTopHat(const SBTopHat& rhs) : SBProfile(rhs) {}
+
+    SBTopHat::~SBTopHat() {}
+
+    double SBTopHat::getRadius() const 
+    {
+        assert(dynamic_cast<const SBTopHatImpl*>(_pimpl.get()));
+        return static_cast<const SBTopHatImpl&>(*_pimpl).getRadius(); 
+    }
+
+    SBTopHat::SBTopHatImpl::SBTopHatImpl(double radius, double flux,
+                                         const GSParamsPtr& gsparams) :
+        SBProfileImpl(gsparams),
+        _r0(radius), _r0sq(_r0*_r0), _flux(flux),
+        _norm(_flux / (M_PI * _r0sq))
+    {
+    }
+
+
+    double SBTopHat::SBTopHatImpl::xValue(const Position<double>& p) const 
+    {
+        double rsq = p.x*p.x + p.y*p.y;
+        if (rsq < _r0sq) return _norm;
+        else return 0.;
+    }
+
+    std::complex<double> SBTopHat::SBTopHatImpl::kValue(const Position<double>& k) const
+    {
+        double kr0sq = (k.x*k.x + k.y*k.y) * _r0sq;
+        return kValue2(kr0sq);
+    }
+
+    std::complex<double> SBTopHat::SBTopHatImpl::kValue2(double kr0sq) const
+    {
+        if (kr0sq < 1.e-4) {
+            // Error from omitting next term is about 1.e-16 for kr0sq = 1.e-4
+            return _flux * (1. - kr0sq * ( (1./8.) + (1./192.) * kr0sq ));
+        } else {
+            double kr0 = sqrt(kr0sq);
+            return 2.*_flux * j1(kr0)/kr0;
+        }
+    }
+
+    void SBTopHat::SBTopHatImpl::fillXValue(tmv::MatrixView<double> val,
+                                            double x0, double dx, int ix_zero,
+                                            double y0, double dy, int iy_zero) const
+    {
+        dbg<<"SBTopHat fillXValue\n";
+        dbg<<"x = "<<x0<<" + ix * "<<dx<<", ix_zero = "<<ix_zero<<std::endl;
+        dbg<<"y = "<<y0<<" + iy * "<<dy<<", iy_zero = "<<iy_zero<<std::endl;
+        if (ix_zero != 0 || iy_zero != 0) {
+            xdbg<<"Use Quadrant\n";
+            fillXValueQuadrant(val,x0,dx,ix_zero,y0,dy,iy_zero);
+        } else {
+            xdbg<<"Non-Quadrant\n";
+            assert(val.stepi() == 1);
+            const int m = val.colsize();
+            const int n = val.rowsize();
+            typedef tmv::VIt<double,1,tmv::NonConj> It;
+
+            val.setZero();
+            for (int j=0;j<n;++j,y0+=dy) {
+                double ysq = y0*y0;
+                double xmax = std::sqrt(_r0sq - ysq);
+                // Set to _norm all pixels with -xmax <= x < xmax
+                // given that x = x0 + i dx.
+                int i1 = std::max(0, int(std::ceil((-xmax - x0)/dx)));
+                int i2 = std::min(m, int(std::ceil((xmax - x0)/dx)));
+                if (i1 < i2)
+                    val.col(j,i1,i2).setAllTo(_norm);
+            }
+        }
+    }
+
+    void SBTopHat::SBTopHatImpl::fillKValue(tmv::MatrixView<std::complex<double> > val,
+                                            double x0, double dx, int ix_zero,
```

Comment:
Yeah, this was a poor design choice I made ages ago, which I just continued here.  I'll take this opportunity to turn them all into kx, ky instead.